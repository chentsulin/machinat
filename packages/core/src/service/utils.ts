import invariant from 'invariant';
import {
  MACHINAT_SERVICE_CONTAINER,
  MACHINAT_SERVICE_PROVIDER,
  MACHINAT_SERVICE_INTERFACE,
} from '../symbol';
import ServiceScope from './scope';
import ServiceMaker from './maker';
import ProvisionMap from './provisionMap';
import type {
  Interfaceable,
  ServiceRequirement,
  ServiceContainer,
  ServiceProvider,
} from './types';

export const isServiceContainer = <T>(
  target: T | ServiceContainer<T>
): target is ServiceContainer<T> =>
  typeof target === 'function' &&
  '$$typeof' in target &&
  target.$$typeof === MACHINAT_SERVICE_CONTAINER;

export const isServiceProvider = (
  target: any
): target is ServiceProvider<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  target.$$typeof === MACHINAT_SERVICE_PROVIDER;

export const maybeInjectContainer = <T>(
  scope: ServiceScope,
  maybeContainer: T | ServiceContainer<T>
): T =>
  isServiceContainer(maybeContainer)
    ? scope.injectContainer(maybeContainer)
    : maybeContainer;

export const isInterfaceable = (target: any): target is Interfaceable<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  (target.$$typeof === MACHINAT_SERVICE_INTERFACE ||
    target.$$typeof === MACHINAT_SERVICE_PROVIDER);

export const createEmptyScope = (platform?: string): ServiceScope =>
  new ServiceScope(new ServiceMaker(new ProvisionMap()), new Map());

/** @internal */
export const polishServiceRequirement = <T>(
  dep: Interfaceable<T> | ServiceRequirement<T>
): ServiceRequirement<T> => {
  if (isInterfaceable(dep)) {
    return { require: dep, optional: false };
  }

  invariant(
    'require' in dep &&
      (dep.require.$$typeof === MACHINAT_SERVICE_INTERFACE ||
        dep.require.$$typeof === MACHINAT_SERVICE_PROVIDER),
    `${(dep as any).name || String(dep)} is not a valid interface`
  );

  return dep;
};
