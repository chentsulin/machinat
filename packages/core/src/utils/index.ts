export { default as compose } from './compose';
export { default as filterSymbolKeys } from './filterSymbolKeys';
export { default as formatNode } from './formatNode';
export * from './isX';
export { default as mixin } from './mixin';
