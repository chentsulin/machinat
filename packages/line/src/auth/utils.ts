/** @internal */ /** */
/* eslint-disable import/prefer-default-export  */
import LineUser from '../user';
import LineChat from '../channel';
import { LIFFAuthData, AuthorizerRefinement } from './types';

export const refinementFromLIFFAuthData = (
  providerId: string,
  botChannelId: string,
  data: LIFFAuthData
): null | AuthorizerRefinement => {
  const { contextType, userId, utouId, groupId, roomId, botChannel } = data;

  let channel: null | LineChat;

  if (botChannel) {
    if (botChannel !== botChannelId || contextType !== 'utou') {
      return null;
    }

    channel = new LineChat(botChannelId, 'utob', userId);
  } else {
    channel =
      contextType === 'utou'
        ? new LineChat(botChannelId, 'utou', utouId as string)
        : contextType === 'room'
        ? new LineChat(botChannelId, 'room', roomId as string)
        : contextType === 'group'
        ? new LineChat(botChannelId, 'group', groupId as string)
        : null;
  }

  return {
    user: new LineUser(providerId, userId),
    channel,
  };
};
