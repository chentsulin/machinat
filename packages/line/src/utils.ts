import { annotateNativeComponent } from '@machinat/core/renderer';
import type { LineSegmentValue, LineMessageSegmentValue } from './types';
import { LINE, CHANNEL_REQUEST_GETTER } from './constant';

export const annotateLineComponent = annotateNativeComponent(LINE);

export const isMessageValue = (
  value: string | LineSegmentValue
): value is string | LineMessageSegmentValue =>
  typeof value === 'string' ||
  (typeof value === 'object' &&
    !Object.prototype.hasOwnProperty.call(value, CHANNEL_REQUEST_GETTER));
