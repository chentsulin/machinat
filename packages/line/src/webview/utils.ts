/** @internal */ /** */
/* eslint-disable import/prefer-default-export  */
import LineUser from '../user';
import LineChat from '../channel';
import { LineAuthContext, AuthorizerRefinement } from './types';

export const refineAuthContext = ({
  contextType,
  userId,
  groupId,
  roomId,
  channelId,
  providerId,
}: LineAuthContext): null | AuthorizerRefinement => ({
  user: new LineUser(providerId, userId),
  channel:
    contextType === 'group'
      ? new LineChat(channelId, 'group', groupId as string)
      : contextType === 'room'
      ? new LineChat(channelId, 'room', roomId as string)
      : new LineChat(channelId, 'user', userId),
});
