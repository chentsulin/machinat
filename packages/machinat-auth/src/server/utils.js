// @flow
import type { IncomingMessage, ServerResponse } from 'http';
import { parse as parseCookie, serialize as serializeCookie } from 'cookie';

export const respondRedirect = (res: ServerResponse, url: string) => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (
  req: IncomingMessage
): null | {| [string]: string |} => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: Object
) => {
  res.setHeader('Set-Cookie', serializeCookie(key, value, options));
};

export const checkDomainScope = (subdomain: string, domain: string) =>
  subdomain.slice(-domain.length) === domain &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const checkPathScope = (subpath: string, path: string) =>
  subpath.slice(0, path.length) === path &&
  (subpath.length === path.length ||
    path[path.length - 1] === '/' ||
    subpath[subpath.length - path.length] === '/');
