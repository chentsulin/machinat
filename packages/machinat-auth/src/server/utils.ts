import type { ServerResponse } from 'http';
import {
  parse as parseCookie,
  serialize as serializeCookie,
  CookieSerializeOptions,
} from 'cookie';
import type { WithHeaders } from '../types';

export const respondRedirect = (res: ServerResponse, url: string): void => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (
  req: WithHeaders
): null | {
  [key: string]: string;
} => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

const SET_COOKIE = 'Set-Cookie';

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: CookieSerializeOptions
): void => {
  const cookieDesc = serializeCookie(key, value, options);
  const cookiesAlreadySet = res.getHeader(SET_COOKIE);

  res.setHeader(
    SET_COOKIE,
    Array.isArray(cookiesAlreadySet)
      ? [...cookiesAlreadySet, cookieDesc]
      : typeof cookiesAlreadySet === 'string'
      ? [cookiesAlreadySet, cookieDesc]
      : cookieDesc
  );
};

export const isSubdomain = (domain: string, subdomain: string): boolean =>
  subdomain.slice(-domain.length) === domain &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const isSubpath = (path: string, subpath: string): boolean =>
  subpath.slice(0, path.length) === path &&
  (subpath.length === path.length ||
    path[path.length - 1] === '/' ||
    subpath[path.length] === '/');
