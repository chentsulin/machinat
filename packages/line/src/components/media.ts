import { makeUnitSegment } from '@machinat/core/renderer';
import { UnitSegment, FunctionOf } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import { LineComponent, LineMessageSegmentValue } from '../types';

/**
 * @category Props
 */
export type VideoProps = {
  /** URL of video file (Max character limit: 1000) */
  originalContentUrl: string;
  /** URL of preview image (Max character limit: 1000) */
  previewImageUrl: string;
  /**
   * ID used to identify the video when Video viewing complete event occurs. If
   * you send a video message with trackingId added, the video viewing complete
   * event occurs when the user finishes watching the video.
   */
  trackingId?: string;
};

/** @internal */
const __Video: FunctionOf<LineComponent<
  VideoProps,
  UnitSegment<LineMessageSegmentValue>
>> = function Video(node, path) {
  const { originalContentUrl, previewImageUrl, trackingId } = node.props;

  return [
    makeUnitSegment(node, path, {
      type: 'video' as const,
      originalContentUrl,
      previewImageUrl,
      trackingId,
    }),
  ];
};
/**
 * Video sends an video message.
 * @category Component
 * @props {@link VideoProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#video-message).
 */
export const Video: LineComponent<
  VideoProps,
  UnitSegment<LineMessageSegmentValue>
> = annotateLineComponent(__Video);

/**
 * @category Props
 */
export type AudioProps = {
  /** URL of audio file (Max character limit: 1000) */
  originalContentUrl: string;
  /** Length of audio file (milliseconds) */
  duration: number;
};

/** @internal */
const __Audio: FunctionOf<LineComponent<
  AudioProps,
  UnitSegment<LineMessageSegmentValue>
>> = function Audio(node, path) {
  const { originalContentUrl, duration } = node.props;
  return [
    makeUnitSegment(node, path, {
      type: 'audio' as const,
      originalContentUrl,
      duration,
    }),
  ];
};
/**
 * Audio sends an audio message.
 * @category Component
 * @props {@link AudioProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#video-message).
 */
export const Audio: LineComponent<
  AudioProps,
  UnitSegment<LineMessageSegmentValue>
> = annotateLineComponent(__Audio);
