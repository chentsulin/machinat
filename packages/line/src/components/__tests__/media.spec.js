import Machinat from '@machinat/core';
import { isNativeType } from '@machinat/core/utils/isX';
import Renderer from '@machinat/core/renderer';

import { Audio, Video } from '../media';

const renderer = new Renderer('line', () => null);

it.each([Audio, Video].map((C) => [C.name, C]))(
  '%s is valid native unit component',
  (_, Media) => {
    expect(typeof Media).toBe('function');

    expect(isNativeType(<Media />)).toBe(true);
    expect(Media.$$platform).toBe('line');
  }
);

it.each(
  [
    <Audio url="https://..." duration={6666} />,
    <Video url="https://..." previewURL="https://..." />,
  ].map((e) => [e.type.name, e])
)('%s render match snapshot', async (_, mediaElement) => {
  const promise = renderer.render(mediaElement);
  await expect(promise).resolves.toEqual([
    {
      type: 'unit',
      node: mediaElement,
      value: expect.any(Object),
      path: '$',
    },
  ]);

  const [{ value }] = await promise;
  expect(value).toMatchSnapshot();
});
