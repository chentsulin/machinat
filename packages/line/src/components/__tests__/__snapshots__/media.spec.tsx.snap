// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio render match snapshot 1`] = `
Object {
  "duration": 6666,
  "originalContentUrl": "https://...",
  "type": "audio",
}
`;

exports[`Video render match snapshot 1`] = `
Object {
  "originalContentUrl": "https://...",
  "previewImageUrl": "https://...",
  "trackingId": undefined,
  "type": "video",
}
`;
